Improve solutions/analysis for Mountain View and Rental Service (#3646)
cpinitiative/usaco-guide: solutions/silver/usaco-787.mdx

id: usaco-787
source: USACO Silver 2018 January
title: Rental Service
author: Óscar Garries, Kevin Sheng, Rishabh Agarwal
---

<Spoiler title="Hint 1">

Remember, at any time Farmer John should either milk the highest-producing cow or rent the lowest-producing cow based on whichever one makes the most money.

</Spoiler>

<Spoiler title="Hint 2">

Farmer John should always sell milk to the store which offers the most money, and rent to the farmer which offers the most money.

> </Spoiler>

Reviewer: these two hints seem to say the same thing?

<Spoiler title="Solution">

[Official Analysis (Java)](http://www.usaco.org/current/data/sol_rental_silver_jan18.html)

## Explanation

To decide whether or not to milk or rent a cow, we should always compare the highest-producing cow with the highest offering store, and the highest offering farmer.

> To do this, we can make 3 lists of stores, farmers, and cows, and sort the stores and farmers by the prices offered, and sort the cows by their milk-production.

Reviewer 2: break this sentence up

> Then, when we go through the list, we either milk the highest-producing cow or rent the lowest producing cow, based on which one makes more money.

Reviewer 2: ```suggestion
Then, when we go through the list, we'll either milk the highest-producing cow or rent the lowest-producing cow, whichever one makes more money.
```
